kafka-connect-hive/src/main/scala/com/landoop/streamreactor/connect/hive/kerberos/KerberosLogin.scala

```diff
@@ -0,0 +1,107 @@
+package com.landoop.streamreactor.connect.hive.kerberos
+
+import java.io.IOException
+import java.net.InetAddress
+import java.security.PrivilegedAction
+
+import scala.concurrent.duration._
+import com.landoop.streamreactor.connect.hive.AsyncFunctionLoop
+import com.typesafe.scalalogging.slf4j.StrictLogging
+import javax.security.auth.login.LoginContext
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.security.SecurityUtil
+import org.apache.hadoop.security.UserGroupInformation
+
+import scala.concurrent.duration.Duration
+
+sealed trait KerberosLogin extends StrictLogging with AutoCloseable {
+  def run[T](thunk: => T):T
+}
+
+case class UserPasswordLogin(ugi: UserGroupInformation, interval: Duration, lc: LoginContext) extends KerberosLogin {
+  private val asyncTicketRenewal = new AsyncFunctionLoop(interval, "Kerberos")(renewKerberosTicket())
+
+  override def close(): Unit = {
+    asyncTicketRenewal.close()
+    lc.logout()
+  }
+
+  private def renewKerberosTicket(): Unit = {
+    try {
+      ugi.reloginFromTicketCache()
+    }
+    catch {
+      case e: IOException =>
+        // We ignore this exception during relogin as each successful relogin gives
+        // additional 24 hours of authentication in the default config. In normal
+        // situations, the probability of failing relogin 24 times is low and if
+        // that happens, the task will fail eventually.
+        logger.error("Error renewing the Kerberos ticket", e)
+    }
+  }
+  override def run[T](thunk: => T): T = {
+    ugi.doAs(new PrivilegedAction[T] {
+      override def run(): T = thunk
+    })
+  }
+}
+
+case class KeytabLogin(ugi: UserGroupInformation, interval: Duration) extends KerberosLogin {
+  private val asyncTicketRenewal = new AsyncFunctionLoop(interval, "Kerberos")(renewKerberosTicket())
```

Comment:
I think you should call start on this unless it's somewhere else

Reply:
done. you were right.